pkg/ruler/rule_backup_manager.go

```diff
@@ -0,0 +1,150 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	promRules "github.com/prometheus/prometheus/rules"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
+)
+
+// Implements GroupLoader interface but instead of reading from a file when Load is called, it returns the
+// rulefmt.RuleGroup it has stored
+type loader struct {
+	ruleGroups map[string][]rulefmt.RuleGroup
+}
+
+func (r *loader) Load(identifier string) (*rulefmt.RuleGroups, []error) {
+	return &rulefmt.RuleGroups{
+		Groups: r.ruleGroups[identifier],
+	}, nil
+}
+
+func (r *loader) Parse(query string) (parser.Expr, error) {
+	return parser.ParseExpr(query)
+}
+
+// rulesBackupManager is an in-memory store that holds []promRules.Group of multiple users. It only stores the Groups,
+// it doesn't evaluate them.
+type rulesBackupManager struct {
+	inMemoryRuleGroupsBackup map[string][]*promRules.Group
+	cfg                      Config
+
+	logger log.Logger
+
+	backupGroupRulesCount      *prometheus.GaugeVec
+	lastBackupReloadSuccessful *prometheus.GaugeVec
+}
+
+func newRulesBackupManager(cfg Config, logger log.Logger, reg prometheus.Registerer) *rulesBackupManager {
+	return &rulesBackupManager{
+		inMemoryRuleGroupsBackup: make(map[string][]*promRules.Group),
+		cfg:                      cfg,
+		logger:                   logger,
+
+		backupGroupRulesCount: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_rule_group_rules",
+			Help:      "The number of backed up rules",
+		}, []string{"user", "rule_group"}),
```

Comment:
I planned for this metric to be the same as `cortex_prometheus_rule_group_rules`. It would give us an understanding on how much data each ruler have as backup. It can also be used to detect discrepancies between what was evaluated and whats on backup which can happen since sync rules happen at different times for each ruler. I don't know when these information could be super useful but I see some potential. if you think the cost of using int over bool is not worth it I will change it :D

Reply:
I still don't get how we gonna use this count. If sync succeeded the backup count should be the same as  `cortex_prometheus_rule_group_rules` for the tenant?

> It can also be used to detect discrepancies between what was evaluated and whats on backup which can happen since sync rules happen at different times for each ruler. I don't know when these information could be super useful but I see some potential.

Yeah, I understand we have such discrepancy but I have no idea how we really gonna use this count. The sync delay is by design. 

I prefer to keep it simple to use boolean or count total number of rules for a user, by removing `rule_group` label

Reply:
What is the conclusion? seems that the `rule_group` is still here.

Reply:
I think I misunderstood it. I kept the `rule_group` label but used boolean. I will update to remove the `rule_group` label and use total count of rules instead.